---
title: "Tutorial of the CausalCellInfer"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Tutorial of the CausalCellInfer}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

In this tutorial, we use the pancreas data to study the type 2 diabetes. The whole analysis took around 10mins to run using our
workstation (Threadripper 7960x, 256GB memory, RTX4070 ti super). 

### Infer the cell markers via the ICP appraoch

In the first step, we need to learn the stable cell markers that will be used for further cell deconvolution. As an example, we used the gender as the environment factor. Theoritically, any
environment that is available can be chosen.

```{r}
library(data.table)
library(stringr)
library(zellkonverter)
library(scater)
library(CausalCellInfer)

sc_dat <- readH5AD("/mnt/md0/yujia/project/github_package/demo/dat/sc_dat/Pancreas/pancreas_marker_identify.h5ad", reader = "R")

# classify the data into 2 different envirs based on diagnosis status
female_index <- which(sc_dat$sex=="female")
male_index <- which(sc_dat$sex=="male")

celltype_all <- as.vector(unlist(sc_dat$cell_type))
unique(celltype_all)

female <- sc_dat[, female_index]
male <- sc_dat[, male_index]

# use xgboost to build a prediction model and extract variable importance for T2D
data_fir <- t(as.matrix(assay(female, "logcounts")))
class(data_fir) <- "numeric"
label_fir <- as.vector(unlist(female$cell_type))

# use xgboost to build a prediction model and extract variable importance for controls
data_sec <- t(as.matrix(assay(male, "logcounts")))
class(data_sec) <- "numeric"
label_sec <- as.vector(unlist(male$cell_type))

# Run the get_cellmarkers to find the marker under the specific environment.
data_1 <- get_cellmarkers(label_fir, data_fir)
data_2 <- get_cellmarkers(label_sec, data_sec)

cellmarkers_common <- list()
cellmarkers_all <- list()

for (i in names(data_1[[2]])){

    cellmakers_fir <- as.data.frame(data_1[[2]][i][[1]])[, 1]
    cellmakers_sec <- as.data.frame(data_2[[2]][i][[1]])[, 1]
    celltype_common <- intersect(cellmakers_fir,cellmakers_sec)
    celltype_all <- union(cellmakers_fir,cellmakers_sec)
    cellmarkers_common[[i]] <- celltype_common
    cellmarkers_all[[i]] <- celltype_all
    # break
}

cellmarkers_all <- unique(unlist(cellmarkers_all))
cellmarkers_all.df <- data.frame("gene" = cellmarkers_all)
dim(cellmarkers_all.df)

cellmarkers_common <- unique(unlist(cellmarkers_common))
cellmarkers_common.df <- data.frame("gene" = cellmarkers_common)
dim(cellmarkers_common.df)

fwrite(cellmarkers_all.df, "/mnt/md0/yujia/project/github_package/demo/dat/marker_genes/Pancreas/all_genes.csv")
fwrite(cellmarkers_common.df, "/mnt/md0/yujia/project/github_package/demo/dat/marker_genes/Pancreas/common_genes.csv")

```

### Estimating the cell composition using the deep learning deconvolution framework.

With the ICP-obtained cell markers, we can estimate the cell composition of the interested bulk RNA-seq dataset with the single-cell RNA-seq dataset as reference.

```{r}
reticulate::use_condaenv("~/miniconda3/envs/causalcellinfer") # We recommend users to set up a seperate conda enviroment to run the analysis.

proportion_estimated <- get_proportion(sc_dat = "/mnt/md0/yujia/project/github_package/demo/dat/sc_dat/Pancreas/diabetes_training.h5ad",
                                       bulk_dat = "/mnt/md0/yujia/project/github_package/demo/dat/bulk/Pancreas/GSE50244_bulk.txt.gz",
                                       marker_genes = "/mnt/md0/yujia/project/github_package/demo/dat/marker_genes/Pancreas/all_genes.csv",
                                       generate_sim_method = "tape", add_noise = FALSE, cut_variance= TRUE, cell_type = "CellType")

proportion_estimated$to_csv("/mnt/md0/yujia/project/github_package/demo/res/Pancreas/proportion/scaden_all.csv")
```

### Infer the CTS profile of different cell types.

After we got the estimated proportion for the given bulk RNA-seq, we can subsequently use ENIGMA approach to infer the CTS profile.

```{r}
CTS_profile <- infer_CTS_profile("/mnt/md0/yujia/project/github_package/demo/dat/bulk/Pancreas/GSE50244_bulk.txt.gz",
                                 "/mnt/md0/yujia/project/github_package/demo/dat/sc_dat/Pancreas/pancreas_training.h5ad",
                                 "/mnt/md0/yujia/project/github_package/demo/res/Pancreas/proportion/scaden_all.csv",
                                 alpha=0.5, beta=0.1, tao_k=0.01, max.iter=1000,
                                 do_cpm = T, model_tracker = F, verbose = T,
                                 model_name = "model_1", Normalize = F, preprocess = "none", pos = F)
```

### Identify the causal gene using PC-simple algorithm

We can further infer the cell-type specific causal gene using the get_causal_gene function.

```{r}
library(CausalCellInfer)
library(data.table)

remove0sd <- function(dataframe){
  dataframe[,which(apply(dataframe, 2, function(x) length(unique(x)))>1)]
}

tmp_cell_exp <- as.data.frame(data.table::fread("/mnt/md0/yujia/project/github_package/demo/res/Pancreas/CTS/alpha_cell.csv.gz")) # this expression profile is estimated using ENIGMA
rownames(tmp_cell_exp) <- tmp_cell_exp$V1
tmp_cell_exp$V1 <- NULL
tmp_cell_exp <- t(tmp_cell_exp)
tmp_cell_exp <- remove0sd(tmp_cell_exp)

meta <- as.data.frame(fread("/mnt/md0/yujia/project/github_package/demo/dat/bulk/Pancreas/GSE50244_meta.csv"))
meta <- meta[!is.na(meta$Hba1c), ]
meta_control <- meta[meta$Hba1c < 6, ]
meta_control$T2D <- 0
meta_case <- meta[meta$Hba1c >= 6.5, ]
meta_case$T2D <- 1
meta <- rbind(meta_control, meta_case)
meta$ID <- paste0("Sample_", meta$ID)

causal_genes <- get_causal_genes(tmp_cell_exp, meta$T2D, 0.01, 10)
```